// rastrojs/dist/enums/types.enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TypesEnum;
(function (TypesEnum) {
    TypesEnum["AA"] = "etiqueta l\u00F3gica sedex";
    TypesEnum["AB"] = "etiqueta l\u00F3gica sedex";
    TypesEnum["AL"] = "agentes de leitura";
    TypesEnum["AR"] = "aviso de recebimento";
    TypesEnum["AS"] = "encomenda pac - acao social";
    TypesEnum["BE"] = "remessa econ\u00F4mica s/ ar digital";
    TypesEnum["BF"] = "remessa expressa s/ ar digital";
    TypesEnum["BG"] = "etiqueta log rem econ c/ar bg";
    TypesEnum["BH"] = "mensagem f\u00EDsico-digital";
    TypesEnum["BI"] = "etiqueta l\u00F3g regist urg";
    TypesEnum["CA"] = "objeto internacional colis";
    TypesEnum["CB"] = "objeto internacional colis";
    TypesEnum["CC"] = "objeto internacional colis";
    TypesEnum["CD"] = "objeto internacional colis";
    TypesEnum["CE"] = "objeto internacional colis";
    TypesEnum["CF"] = "objeto internacional colis";
    TypesEnum["CG"] = "objeto internacional colis";
    TypesEnum["CH"] = "objeto internacional colis";
    TypesEnum["CI"] = "objeto internacional colis";
    TypesEnum["CJ"] = "objeto internacional colis";
    TypesEnum["CK"] = "objeto internacional colis";
    TypesEnum["CL"] = "objeto internacional colis";
    TypesEnum["CM"] = "objeto internacional colis";
    TypesEnum["CN"] = "objeto internacional colis";
    TypesEnum["CO"] = "objeto internacional colis";
    TypesEnum["CP"] = "objeto internacional colis";
    TypesEnum["CQ"] = "objeto internacional colis";
    TypesEnum["CR"] = "objeto internacional colis";
    TypesEnum["CS"] = "objeto internacional colis";
    TypesEnum["CT"] = "objeto internacional colis";
    TypesEnum["CU"] = "objeto internacional colis";
    TypesEnum["CV"] = "objeto internacional colis";
    TypesEnum["CW"] = "objeto internacional colis";
    TypesEnum["CX"] = "objeto internacional colis";
    TypesEnum["CY"] = "objeto internacional colis";
    TypesEnum["CZ"] = "objeto internacional colis";
    TypesEnum["DA"] = "encomenda sedex c/ ar digital";
    TypesEnum["DB"] = "remessa expressa c/ ar digital-bradesco";
    TypesEnum["DC"] = "remessa expressa (orgao transito)";
    TypesEnum["DD"] = "devolu\u00E7\u00E3o de documentos";
    TypesEnum["DE"] = "remessa expressa c/ ar digital";
    TypesEnum["DF"] = "encomenda sedex (etiq logica)";
    TypesEnum["DG"] = "encomenda sedex (etiq logica)";
    TypesEnum["DI"] = "rem expres com ar digital itau";
    TypesEnum["DJ"] = "encomenda sedex";
    TypesEnum["DK"] = "sedex extra grande";
    TypesEnum["DL"] = "sedex l\u00F3gico";
    TypesEnum["DM"] = "encomenda sedex";
    TypesEnum["DN"] = "encomenda sedex";
    TypesEnum["DO"] = "rem expres com ar digital itau unibanco";
    TypesEnum["DP"] = "sedex pagamento entrega";
    TypesEnum["DQ"] = "rem expres com ar digital bradesco";
    TypesEnum["DR"] = "rem expres com ar digital santander";
    TypesEnum["DS"] = "rem expres com ar digital santander";
    TypesEnum["DT"] = "remessa econ.seg.transito c/ar digital";
    TypesEnum["DU"] = "encomenda sedex";
    TypesEnum["DV"] = "sedex com ar digital";
    TypesEnum["DW"] = "encomenda sedex (etiq l\u00F3gica)";
    TypesEnum["DX"] = "sedex 10 l\u00F3gico";
    TypesEnum["DY"] = "encomenda sedex (etiq f\u00EDsica)";
    TypesEnum["DZ"] = "encomenda sedex (etiq logica)";
    TypesEnum["EA"] = "objeto internacional ems";
    TypesEnum["EB"] = "objeto internacional ems";
    TypesEnum["EC"] = "encomenda pac";
    TypesEnum["ED"] = "objeto internacional packet express";
    TypesEnum["EE"] = "objeto internacional ems";
    TypesEnum["EF"] = "objeto internacional ems";
    TypesEnum["EG"] = "objeto internacional ems";
    TypesEnum["EH"] = "objeto internacional ems";
    TypesEnum["EI"] = "objeto internacional ems";
    TypesEnum["EJ"] = "objeto internacional ems";
    TypesEnum["EK"] = "objeto internacional ems";
    TypesEnum["EL"] = "objeto internacional ems";
    TypesEnum["EM"] = "sedex mundi";
    TypesEnum["EN"] = "objeto internacional ems";
    TypesEnum["EO"] = "objeto internacional ems";
    TypesEnum["EP"] = "objeto internacional ems";
    TypesEnum["EQ"] = "encomenda servi\u00E7o n\u00E3o expressa ect";
    TypesEnum["ER"] = "registrado";
    TypesEnum["ES"] = "objeto internacional ems";
    TypesEnum["ET"] = "objeto internacional ems";
    TypesEnum["EU"] = "objeto internacional ems";
    TypesEnum["EV"] = "objeto internacional ems";
    TypesEnum["EW"] = "objeto internacional ems";
    TypesEnum["EX"] = "objeto internacional ems";
    TypesEnum["EY"] = "objeto internacional ems";
    TypesEnum["EZ"] = "objeto internacional ems";
    TypesEnum["FA"] = "fac registrado";
    TypesEnum["FB"] = "fac registrado";
    TypesEnum["FC"] = "fac registrado (5 dias)";
    TypesEnum["FD"] = "fac registrado (10 dias)";
    TypesEnum["FE"] = "encomenda fnde";
    TypesEnum["FF"] = "registrado detran";
    TypesEnum["FH"] = "fac registrado c/ ar digital";
    TypesEnum["FJ"] = "remessa econ\u00F4mica c/ ar digital";
    TypesEnum["FM"] = "fac registrado (monitorado)";
    TypesEnum["FR"] = "fac registrado";
    TypesEnum["IA"] = "integrada avulsa";
    TypesEnum["IC"] = "integrada a cobrar";
    TypesEnum["ID"] = "integrada devolucao de documento";
    TypesEnum["IE"] = "integrada especial";
    TypesEnum["IF"] = "cpf";
    TypesEnum["II"] = "integrada interno";
    TypesEnum["IK"] = "integrada com coleta simultanea";
    TypesEnum["IM"] = "integrada medicamentos";
    TypesEnum["IN"] = "obj de corresp e ems rec exterior";
    TypesEnum["IP"] = "integrada programada";
    TypesEnum["IR"] = "impresso registrado";
    TypesEnum["IS"] = "integrada standard";
    TypesEnum["IT"] = "integrada termol\u00E1bil";
    TypesEnum["IU"] = "integrada urgente";
    TypesEnum["IX"] = "edei encomenda expressa";
    TypesEnum["JA"] = "remessa economica c/ar digital";
    TypesEnum["JB"] = "remessa economica c/ar digital";
    TypesEnum["JC"] = "remessa economica c/ar digital";
    TypesEnum["JD"] = "remessa economica s/ar digital";
    TypesEnum["JE"] = "remessa economica c/ar digital";
    TypesEnum["JF"] = "remessa economica c/ar digital";
    TypesEnum["JG"] = "registrado priorit\u00E1rio";
    TypesEnum["JH"] = "registrado priorit\u00E1rio";
    TypesEnum["JI"] = "remessa economica s/ar digital";
    TypesEnum["JJ"] = "registrado justi\u00E7a";
    TypesEnum["JK"] = "remessa econ\u00F4mica s/ar digital";
    TypesEnum["JL"] = "registrado l\u00F3gico";
    TypesEnum["JM"] = "mala direta postal especial";
    TypesEnum["JN"] = "mala direta postal especial";
    TypesEnum["JO"] = "registrado priorit\u00E1rio";
    TypesEnum["JP"] = "objeto receita federal (exclusivo)";
    TypesEnum["JQ"] = "remessa economica c/ar digital";
    TypesEnum["JR"] = "registrado priorit\u00E1rio";
    TypesEnum["JS"] = "registrado l\u00F3gico";
    TypesEnum["JT"] = "registrado urgente";
    TypesEnum["JU"] = "etiqueta f\u00EDs regist urg";
    TypesEnum["JV"] = "remessa econ\u00F4mica c/ar digital";
    TypesEnum["JW"] = "carta comercial a faturar (5 dias)";
    TypesEnum["JX"] = "carta comercial a faturar (10 dias)";
    TypesEnum["JY"] = "remessa economica (5 dias)";
    TypesEnum["JZ"] = "remessa economica (10 dias)";
    TypesEnum["LA"] = "log\u00EDstica reversa simult\u00E2nea sedex";
    TypesEnum["LB"] = "log\u00EDstica reversa simult\u00E2nea sedex";
    TypesEnum["LC"] = "objeto internacional prime";
    TypesEnum["LD"] = "objeto internacional prime";
    TypesEnum["LE"] = "log\u00EDstica reversa economica";
    TypesEnum["LF"] = "objeto internacional prime";
    TypesEnum["LG"] = "objeto internacional prime";
    TypesEnum["LH"] = "objeto internacional prime";
    TypesEnum["LI"] = "objeto internacional prime";
    TypesEnum["LJ"] = "objeto internacional prime";
    TypesEnum["LK"] = "objeto internacional prime";
    TypesEnum["LL"] = "objeto internacional prime";
    TypesEnum["LM"] = "objeto internacional prime";
    TypesEnum["LN"] = "objeto internacional prime";
    TypesEnum["LP"] = "log\u00EDstica reversa simult\u00E2nea pac";
    TypesEnum["LQ"] = "objeto internacional prime";
    TypesEnum["LS"] = "logistica reversa sedex";
    TypesEnum["LV"] = "logistica reversa expressa";
    TypesEnum["LW"] = "objeto internacional prime";
    TypesEnum["LX"] = "objeto internacional packet economic";
    TypesEnum["LY"] = "objeto internacional prime";
    TypesEnum["LZ"] = "objeto internacional prime";
    TypesEnum["MA"] = "telegrama - servicos adicionais";
    TypesEnum["MB"] = "telegrama de balcao";
    TypesEnum["MC"] = "telegrama fonado";
    TypesEnum["MD"] = "maquina de franquear (logica)";
    TypesEnum["ME"] = "telegrama";
    TypesEnum["MF"] = "telegrama fonado";
    TypesEnum["MH"] = "carta via internet";
    TypesEnum["MK"] = "telegrama corporativo";
    TypesEnum["MM"] = "telegrama grandes clientes";
    TypesEnum["MP"] = "telegrama pr\u00E9-pago";
    TypesEnum["MS"] = "encomenda saude";
    TypesEnum["MT"] = "telegrama via telemail";
    TypesEnum["MY"] = "telegrama internacional entrante";
    TypesEnum["MZ"] = "telegrama via correios on line";
    TypesEnum["NE"] = "tele sena resgatada";
    TypesEnum["NX"] = "edei encomenda nao urgente";
    TypesEnum["OA"] = "encomenda sedex (etiq logica)";
    TypesEnum["OB"] = "encomenda sedex (etiq logica)";
    TypesEnum["OC"] = "encomenda sedex (etiq logica)";
    TypesEnum["OD"] = "encomenda sedex (etiq f\u00EDsica)";
    TypesEnum["OF"] = "etiqueta l\u00F3gica sedex";
    TypesEnum["OG"] = "etiqueta l\u00F3gica sedex";
    TypesEnum["OH"] = "etiqueta l\u00F3gica sedex";
    TypesEnum["PA"] = "passaporte";
    TypesEnum["PB"] = "encomenda pac - n\u00E3o urgente";
    TypesEnum["PC"] = "encomenda pac a cobrar";
    TypesEnum["PD"] = "encomenda pac";
    TypesEnum["PE"] = "encomenda pac (etiqueta fisica)";
    TypesEnum["PF"] = "passaporte";
    TypesEnum["PG"] = "encomenda pac (etiqueta fisica)";
    TypesEnum["PH"] = "encomenda pac (etiqueta logica)";
    TypesEnum["PI"] = "encomenda pac";
    TypesEnum["PJ"] = "encomenda pac";
    TypesEnum["PK"] = "pac extra grande";
    TypesEnum["PL"] = "encomenda pac";
    TypesEnum["PM"] = "encomenda pac (etiq f\u00EDsica)";
    TypesEnum["PN"] = "encomenda pac (etiq logica)";
    TypesEnum["PO"] = "encomenda pac (etiq logica)";
    TypesEnum["PP"] = "etiqueta l\u00F3gica pac";
    TypesEnum["PQ"] = "etiqueta logica pac mini";
    TypesEnum["PR"] = "reembolso postal - cliente avulso";
    TypesEnum["PS"] = "etiqueta l\u00F3gica pac";
    TypesEnum["PT"] = "etiqueta l\u00F3gica pac";
    TypesEnum["PU"] = "etiqueta l\u00F3gica pac";
    TypesEnum["PV"] = "etiqueta l\u00F3gica pac administrativo";
    TypesEnum["PX"] = "etiqueta l\u00F3gica pac";
    TypesEnum["RA"] = "registrado priorit\u00E1rio";
    TypesEnum["RB"] = "carta registrada";
    TypesEnum["RC"] = "carta registrada com valor declarado";
    TypesEnum["RD"] = "remessa economica detran";
    TypesEnum["RE"] = "mala direta postal especial";
    TypesEnum["RF"] = "objeto da receita federal";
    TypesEnum["RG"] = "registrado do sistema sara";
    TypesEnum["RH"] = "registrado com ar digital";
    TypesEnum["RI"] = "registrado priorit\u00E1rio internacional";
    TypesEnum["RJ"] = "registrado ag\u00EAncia";
    TypesEnum["RK"] = "registrado ag\u00EAncia";
    TypesEnum["RL"] = "registrado l\u00F3gico";
    TypesEnum["RM"] = "registrado ag\u00EAncia";
    TypesEnum["RN"] = "registrado ag\u00EAncia";
    TypesEnum["RO"] = "registrado ag\u00EAncia";
    TypesEnum["RP"] = "reembolso postal - cliente inscrito";
    TypesEnum["RQ"] = "registrado ag\u00EAncia";
    TypesEnum["RR"] = "registrado internacional";
    TypesEnum["RS"] = "rem econ org transito com ou sem ar";
    TypesEnum["RT"] = "rem econ talao/cartao sem ar digita";
    TypesEnum["RU"] = "registrado servi\u00E7o ect";
    TypesEnum["RV"] = "rem econ crlv/crv/cnh com ar digital";
    TypesEnum["RW"] = "registrado internacional";
    TypesEnum["RX"] = "registrado internacional";
    TypesEnum["RY"] = "rem econ talao/cartao com ar digital";
    TypesEnum["RZ"] = "registrado";
    TypesEnum["SA"] = "etiqueta sedex ag\u00EAncia";
    TypesEnum["SB"] = "sedex 10";
    TypesEnum["SC"] = "sedex a cobrar";
    TypesEnum["SD"] = "remessa expressa detran";
    TypesEnum["SE"] = "encomenda sedex";
    TypesEnum["SF"] = "sedex agencia";
    TypesEnum["SG"] = "sedex do sistema sara";
    TypesEnum["SH"] = "sedex com ar digital";
    TypesEnum["SI"] = "sedex ag\u00EAncia";
    TypesEnum["SJ"] = "sedex hoje";
    TypesEnum["SK"] = "sedex ag\u00EAncia";
    TypesEnum["SL"] = "sedex l\u00F3gico";
    TypesEnum["SM"] = "sedex 12";
    TypesEnum["SN"] = "sedex ag\u00EAncia";
    TypesEnum["SO"] = "sedex ag\u00EAncia";
    TypesEnum["SP"] = "sedex pr\u00E9-franqueado";
    TypesEnum["SQ"] = "sedex";
    TypesEnum["SR"] = "sedex";
    TypesEnum["SS"] = "sedex f\u00EDsico";
    TypesEnum["ST"] = "rem expres talao/cartao sem ar digital";
    TypesEnum["SU"] = "encomenda servi\u00E7o expressa ect";
    TypesEnum["SV"] = "rem expres crlv/crv/cnh com ar digital";
    TypesEnum["SW"] = "encomenda sedex";
    TypesEnum["SX"] = "sedex 10";
    TypesEnum["SY"] = "rem expres talao/cartao com ar digital";
    TypesEnum["SZ"] = "sedex ag\u00EAncia";
    TypesEnum["TC"] = "teste (objeto para treinamento)";
    TypesEnum["TE"] = "teste (objeto para treinamento)";
    TypesEnum["TR"] = "objeto treinamento - n\u00E3o gera pr\u00E9-alerta";
    TypesEnum["TS"] = "teste (objeto para treinamento)";
    TypesEnum["VA"] = "objeto internacional com valor declarado";
    TypesEnum["VC"] = "objeto internacional com valor declarado";
    TypesEnum["VD"] = "objeto internacional com valor declarado";
    TypesEnum["VE"] = "objeto internacional com valor declarado";
    TypesEnum["VF"] = "objeto internacional com valor declarado";
    TypesEnum["VV"] = "objeto internacional com valor declarado";
    TypesEnum["XA"] = "aviso chegada objeto int tributado";
    TypesEnum["XM"] = "sedex mundi";
    TypesEnum["XR"] = "objeto internacional (pps tributado)";
    TypesEnum["XX"] = "objeto internacional (pps tributado)";
})(TypesEnum = exports.TypesEnum || (exports.TypesEnum = {}));
